src/pages/SignUp.tsx

```diff
@@ -0,0 +1,127 @@
+import React, { useContext, useState } from "react";
+import { Link } from "react-router-dom";
+import { useHistory } from "react-router";
+import { ValidationWarning } from "../components";
+import { useAuth } from "../contexts/auth-context/authContext";
+
+const SignUp = () => {
+  const { signUpUser } = useAuth();
+  const [password, setPassword] = useState("");
+  const [firstName, setFirstName] = useState("");
+  const [lastName, setLastName] = useState("");
+  const [email, setEmail] = useState("");
+  const [isEmailValid, setIsEmailValid] = useState(false);
+  const [isPasswordValid, setIsPasswordValid] = useState(false);
+  const [isFirstNameValid, setIsFirstNameValid] = useState(false);
+  const [isLastNameValid, setIsLastNameValid] = useState(false);
+  const [isValid, setIsValid] = useState(true);
+
+  let history = useHistory();
+
+  const handleSignUpClick = () => {
+    if (
+      isFirstNameValid &&
+      isLastNameValid &&
+      isPasswordValid &&
+      isEmailValid
+    ) {
+      setIsValid(true);
+      signUpUser(firstName, lastName, email, password);
+      history.push("/signin");
+    } else {
+      setIsValid(false);
+    }
+  };
+
+  const emailChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setEmail(e.target.value);
+    const emailRE =
+      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+
+    emailRE.test(String(e.target.value).toLowerCase())
+      ? setIsEmailValid(true)
+      : setIsEmailValid(false);
+  };
+
+  const FirstNameChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setFirstName(e.target.value);
+    e.target.value === ""
+      ? setIsFirstNameValid(false)
+      : setIsFirstNameValid(true);
+  };
+  const LastNameChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setLastName(e.target.value);
+    e.target.value === ""
+      ? setIsLastNameValid(false)
+      : setIsLastNameValid(true);
+  };
+
+  const passwordChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+    e.target.value === ""
+      ? setIsPasswordValid(false)
+      : setIsPasswordValid(true);
+  };
+
+  return (
+    <div className="text-center mt-8">
+      <h1 className=" text-3xl mb-4 font-semibold">Sign Up to RQUiz</h1>
+      <form action="">
+        <div className="mt-12">
+          <div className="mb-3 pt-0 my-4">
```

Comment:
same div for input is used with same style so the handler can be added as config but instead of having same div repeated again, you can use map function to map over the config array and display that in div